[SYCL] fix set main gpu error, support single/mul gpu mode

NeoZhangJianyu opened the pull request:
1. fix error to set main gpu as non-zero.
2. add new APIs: ggml_backend_sycl_set_single_device_mode(), ggml_backend_sycl_set_mul_device_mode() to handle single/multiple cards by split-mode.
  In CI, enable ggml_backend_sycl_set_mul_device_mode() as default.
  If split-mode==layer (default), use all GPUs with top max compute unit.
  Else, use the main-gpu set by user as only device. It supports level_zero:gpu, opencl:gpu.

3. refactor the shown device list.
 - change the order in shown device list.
   sort the devices by type (level_zero, opencl:gpu, opencl:cpu, opencl:acc) and max compute unit.
 - add type item to follow the output style of tool **sycl-ls**.
```
found 6 SYCL devices:
|  |                  |                                             |compute   |Max compute|Max work|Max sub|               |
|ID|       Device Type|                                         Name|capability|units      |group   |group  |Global mem size|
|--|------------------|---------------------------------------------|----------|-----------|--------|-------|---------------|
| 0|[level_zero:gpu:0]|               Intel(R) Arc(TM) A770 Graphics|       1.3|        512|    1024|     32|    16225243136|
| 1|[level_zero:gpu:1]|                    Intel(R) UHD Graphics 770|       1.3|         32|     512|     32|    53651849216|
| 2|    [opencl:gpu:0]|               Intel(R) Arc(TM) A770 Graphics|       3.0|        512|    1024|     32|    16225243136|
| 3|    [opencl:gpu:1]|                    Intel(R) UHD Graphics 770|       3.0|         32|     512|     32|    53651849216|
| 4|    [opencl:cpu:0]|         13th Gen Intel(R) Core(TM) i7-13700K|       3.0|         24|    8192|     64|    67064815616|
| 5|    [opencl:acc:0]|               Intel(R) FPGA Emulation Device|       1.2|         24|67108864|     64|    67064815616|
```

4. update examples/sycl/run-llama2.sh
   support to switch in single/multiple cards by set device id as parameter.

NeoZhangJianyu committed:
fix set main gpu error, support single/mul gpu mode

NeoZhangJianyu committed:
update format

slaren committed:
ci : remove tidy-review (#6021)

slaren commented:
The same concerns that I mentioned in the first review are still present here. My recommendation is to remove `ggml_backend_sycl_set_single_device_mode`, `ggml_backend_sycl_get_device_index`, `ggml_backend_sycl_set_mul_device_mode` entirely and figure a way to solve this problem that doesn't require adding more backend-specific functions.

NeoZhangJianyu commented:
> The same concerns that I mentioned in the first review are still present here. My recommendation is to remove `ggml_backend_sycl_set_single_device_mode`, `ggml_backend_sycl_get_device_index`, `ggml_backend_sycl_set_mul_device_mode` entirely and figure a way to solve this problem that doesn't require adding more backend-specific functions.

Invite @0cc4m, @airMeng, @abhilash1910 @AidanBeltonS 
  
SYCL backend could support different device types, like iGPU/dGPU, CPU, FPGA (which is like Vulkan).
It needs to support them and special cases. 
In current common framework, we only could modify SYCL code part. That limits to provide better solution.
This PR provides a new method to support new cases with less impact to common part.
If llama.cpp could provide better solution in architecture, SYCL could use the common solution.

In current framework, following cases are supported:
1.  1 dGPU or 1 iGPU
2.  2+dGPU

But SYCL need to support the more cases: 
3. 1 iGPU + 1 dGPU
4. 1 iGPU + 2+ dGPU
User could choose one of iGPU/dGPU in case 3, 4, or choose 2+ dGPU in case 4.

This PR avoid mixing iGPU and dGPU together, that will reduce the performance obviously. But current framework allows it.
In multiple cards case, SYCL will choose all GPUs with same power, instead of mix all GPUs.
In single card case, SYCL could select iGPU or dGPU. 

To cover case 1, 2, I think SYCL could follow current framework.
But for case 3,4, there is no existed solution.
That's why we try new solution/GGML APIs for them.

In the future, we want SYCL to support CPU (case 5) to cover CI/unit-test requirement, so that make sure the SYCL backend don't be broken in new PR.
5. CPU

More cases in the future:
6. iGPU + CPU + FPGA
7. iGPU + CPU + dGPU
mix different devices or not mix.

I guess other backends have similar requirement to support more device types.
We hope llama.cpp provide a common solution, so that we could follow it to reduce the cost.

Thank you!

slaren commented:
How could we improve the llama.cpp framework to support these cases better?

Something I would like to do in the future is remove most of the backend-specific initialization code in llama.cpp and use the ggml-backend registry in a generic way. Then the user could specify the devices that they want to use by name. For example, the user could specify to use devices `cpu`, `sycl_igpu0` and `sycl_dgpu0` to select CPU, iGPU and dGPU.

NeoZhangJianyu commented:
Yes, I think it's good solution.

When will you plan to implement it?

Could SYCL backend wait for this new solution? or fix the issue of set main gpu by this PR and update when common solution is ready?

abhilash1910 reviewed (approved):
LGTM! I think it is good to merge this PR since this fixes breaking issue , we can later modify based on generic ggml backend initialization.

ngxson committed:
Server: Use multi-task for embeddings endpoint (#6001)

* use multitask for embd endpoint

* specify types

* remove redundant {"n_predict", 0}

slaren commented:
> Could SYCL backend wait for this new solution? or fix the issue of set main gpu by this PR and update when common solution is ready?

It's going to take a while, it's not a priority at the moment. It is ok to use a temporary solution for now.

AidanBeltonS commented on ggml-sycl.cpp:
```diff
@@ -937,17 +961,65 @@ namespace dpct
+            dpct::get_device_info(prop2, device2);
+            return prop1.get_max_compute_units() > prop2.get_max_compute_units();
+        }
+        static int convert_backend_index(std::string & backend) {
+            if (backend == "ext_oneapi_level_zero:gpu") return 0;
+            if (backend == "opencl:gpu") return 1;
+            if (backend == "opencl:cpu") return 2;
+            if (backend == "opencl:acc") return 3;
+            printf("convert_backend_index: can't handle backend=%s\n", backend.c_str());
+            GGML_ASSERT(false);
+        }
```
Current approach fails on NVidia and AMD targets
```suggestion
        static int convert_backend_index(std::string & backend) {
            if (backend == "ext_oneapi_level_zero:gpu") return 0;
            if (backend == "opencl:gpu") return 1;
            if (backend == "opencl:cpu") return 2;
            if (backend == "opencl:acc") return 3;
            if (backend == "ext_oneapi_cuda:gpu") return 4;
            if (backend == "ext_oneapi_hip:gpu") return 5;
            printf("convert_backend_index: can't handle backend=%s\n", backend.c_str());
            GGML_ASSERT(false);
        }
```

abhilash1910 replied:
Yes this is because of the gpu branches for amd/nv were not handled yet.

NeoZhangJianyu replied:
I want to focus to support Intel GPU only in this PR.
I can't make sure the above code work well with other vendor GPUs, because I have no hardware environment to test other vendor GPUs.

I suggest you create new PR based on this PR to support other vendor GPUs.

Here, the index impacts the order in device list. lower is for higher priority.
I suggest change to 2,3, move CPU and ACC to 4, 5 in new PR.

AidanBeltonS reviewed (commented):
One necessary fix, and two small NITs otherwise it looks fine.

AidanBeltonS committed:
Update get version (#6025)

slaren committed:
test-backend-ops : skip CPU backend by default (#6028)

slaren committed:
llama : add pipeline parallelism support (#6017)

* llama : add pipeline parallelism support for batch processing with multiple CUDA GPUs

ggml-ci

* server : add -ub, --ubatch-size parameter

* fix server embedding test

* llama : fix Mamba inference for pipeline parallelism

Tested to work correctly with both `main` and `parallel` examples.

* llama : limit max batch size to n_batch

* add LLAMA_SCHED_MAX_COPIES to configure the number of input copies for pipeline parallelism
default increase to 4 (from 2)

changing this value may improve performance for some systems, but increases memory usage

* fix hip build

* fix sycl build (disable cpy_tensor_async)

* fix hip build

* llama : limit n_batch and n_ubatch to n_ctx during context creation

* llama : fix norm backend

* batched-bench : sync after decode

* swiftui : sync after decode

* ggml : allow ggml_get_rows to use multiple threads if they are available

* check n_ubatch >= n_tokens with non-casual attention

* llama : do not limit n_batch to n_ctx with non-casual attn

* server : construct batch with size of llama_n_batch

* ggml_backend_cpu_graph_compute : fix return value when alloc fails

* llama : better n_batch and n_ubatch comment

* fix merge

* small fix

* reduce default n_batch to 2048

---------

Co-authored-by: Francis Couture-Harpin <git@compilade.net>
Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

HanClinto committed:
grammar : handle missing "root" node (#6004)

ggerganov committed:
readme : update API changes and hot topics

wanix1988 committed:
readme : update details about running llama in Termux on Android (#6039)

NeoZhangJianyu committed:
fix grammar issue

Co-authored-by: AidanBeltonS <87009434+AidanBeltonS@users.noreply.github.com>

NeoZhangJianyu committed:
fix grammar issue

Co-authored-by: AidanBeltonS <87009434+AidanBeltonS@users.noreply.github.com>

NeoZhangJianyu commented:
> > Could SYCL backend wait for this new solution? or fix the issue of set main gpu by this PR and update when common solution is ready?
> 
> It's going to take a while, it's not a priority at the moment. It is ok to use a temporary solution for now.

Got it! Thank you!

ggerganov committed:
embedding : print cosine similarity (#899)

ggerganov committed:
metal : build metallib + fix embed path (#6015)

* metal : build metallib + fix embed path

ggml-ci

* metal : fix embed build + update library load logic

ggml-ci

* metal : fix embeded library build

ggml-ci

* ci : fix iOS builds to use embedded library

ggerganov committed:
embedding : print all resulting embeddings (#899)

ggerganov committed:
ggml : designate enum vals for integer types (#6050)

certik committed:
gguf-py : add support for I8, I16 and I32 (#6045)

* Refactor dtype handling to be extensible

This code is equivalent as before, but now it is prepared to easily add
more NumPy dtypes.

* Add support for I8, I16 and I32

These types are allowed in the GGUF specification.

* Add support for I8, I16 and I32 to gguf_writer

* Add support for I8, I16, I32 to gguf_reader

Xarbirus committed:
llama : optimize defrag moves + fix fragmentation calculation (#6037)

* attempt to reduce the impact of a worst-case scenario

* fragmentation calculation fix

* Update llama.cpp

---------

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ggerganov committed:
llama : fix typo

phymbert committed:
server: disable debug release type sanitizer, simplify trigger (#6047)

- increase time out for server
 - do not fail fast

jianliao committed:
readme : improve readme for Llava-1.6 example (#6044)

Co-authored-by: Jian Liao <jianliao@adobe.com>

ggerganov committed:
gguf-py : fix dtype check (#6045)

NeoZhangJianyu commented:
@ggerganov Could you review this PR?

ggerganov commented on llama.cpp:
```diff
@@ -5055,6 +5055,16 @@ static int llama_model_load(const std::string & fname, llama_model & model, llam
         }
 #endif
 
+#ifdef GGML_USE_SYCL
+    if (params.split_mode == LLAMA_SPLIT_MODE_NONE) {
+        ggml_backend_sycl_set_single_device_mode(params.main_gpu);
+        //SYCL use device index (0, 1, 2) directly, uer input device id, then convert to device index.
+        params.main_gpu = ggml_backend_sycl_get_device_index(params.main_gpu);
+    } else {
+        ggml_backend_sycl_set_mul_device_mode();
+    }
+#endif
+
```
Fix indentation

NeoZhangJianyu replied:
yes, fix it

ggerganov commented on llama.cpp:
```diff
@@ -12828,23 +12838,22 @@ struct llama_context * llama_new_context_with_model(
                 if (backend == nullptr) {
-                    LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d)backend\n", __func__, model->main_gpu, main_gpu_index);
+                    int main_gpu_id = ggml_backend_sycl_get_device_id(model->main_gpu);
+                    LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d)backend\n", __func__, main_gpu_id, model->main_gpu);
```
```suggestion
                    LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d) backend\n", __func__, main_gpu_id, model->main_gpu);
```

ggerganov commented on llama.cpp:
```diff
@@ -12828,23 +12838,22 @@ struct llama_context * llama_new_context_with_model(
-                        LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d)backend\n", __func__, device_id, i);
+                        int id_list[GGML_SYCL_MAX_DEVICES];
+                        ggml_sycl_get_gpu_list(id_list, GGML_SYCL_MAX_DEVICES);
+                        LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d)backend\n", __func__, id_list[i], i);
```
```suggestion
                        LLAMA_LOG_ERROR("%s: failed to initialize SYCL%d (index %d) backend\n", __func__, id_list[i], i);
```

ggerganov commented on ggml-sycl.h:
```diff
@@ -28,7 +28,9 @@ GGML_API GGML_CALL int   ggml_backend_sycl_get_device_count();
 GGML_API GGML_CALL void ggml_backend_sycl_get_device_memory(int device, size_t *free, size_t *total);
 GGML_API GGML_CALL int ggml_backend_sycl_get_device_index(int device_id);
-
+GGML_API GGML_CALL int ggml_backend_sycl_get_device_id(int device_index);
+GGML_API GGML_CALL void ggml_backend_sycl_set_single_device_mode(int main_gpu_id);
+GGML_API GGML_CALL void ggml_backend_sycl_set_mul_device_mode();
```
```suggestion

// TODO: these are temporary
//       ref: https://github.com/ggerganov/llama.cpp/pull/6022#issuecomment-1992615670
GGML_API GGML_CALL int ggml_backend_sycl_get_device_id(int device_index);
GGML_API GGML_CALL void ggml_backend_sycl_set_single_device_mode(int main_gpu_id);
GGML_API GGML_CALL void ggml_backend_sycl_set_mul_device_mode();
```

NeoZhangJianyu committed:
fix format

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

NeoZhangJianyu committed:
fix format

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

NeoZhangJianyu committed:
mark comment for improvement in the future

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

NeoZhangJianyu committed:
fix format

ggerganov committed:
embedding : add EOS token if not present (#899)

Xarbirus committed:
llama : support models without vocabulary (#5798)

* additional methods to read model and ctx parameters

* vocab size as a part of a model metadata

* models without vocabulary, convert.py part

* models without vocabulary, llama.cpp part

* PR clean up

* converter scrypt fixes

* llama_vocab_type update (renamed the new key)

* pr review fixes

* revert function renaming

* one more NoVocab assert

certik committed:
gguf-py : bump version to 0.8.0 (#6060)

stevegrubb committed:
gguf : fix resource leaks (#6061)

There several places where a gguf context is allocated. A call to gguf_free
is missing in some error paths. Also on linux, llama-bench was missing a
fclose.

ggerganov committed:
llama : fix integer overflow during quantization (#6063)

NeoZhangJianyu commented:
@ggerganov Some CI cases have be passed. But now they are always fail.
The PR won't impact the fault cases in fact.
Could I merge the code directly or re-try the fault CI cases until they are passed?

Thank you!

NeoZhangJianyu commented:
Close it to rebase and submit again

zhouwg commented:
> How could we improve the llama.cpp framework to support these cases better?
> 
> Something I would like to do in the future is remove most of the backend-specific initialization code in llama.cpp and use the ggml-backend registry in a generic way. Then the user could specify the devices that they want to use by name. For example, the user could specify to use devices `cpu`, `sycl_igpu0` and `sycl_dgpu0` to select CPU, iGPU and dGPU.

I agree with your idea:remove most of the backend-specific initialization code in llama.cpp and use the ggml-backend registry in a generic way

NeoZhangJianyu commented:
Yes. This refactor is ongoing. but need more time.
